[2.9.0] Document the SQLite-backed caching experimental feature (#1337)
rancher/rancher-docs: docs/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching.md

---

> title: SQLite-Backed Caching

Reviewer: ```suggestion
title: SQLite-Backed Caching and UI Server-Side Pagination
```

Wonder if we need to make this a little clearer i.e backend cache changes and frontend changes that make use of it. I've tweaked a few things below, but could we call this out in the title?

Author: From a user perspective I actually think only UI is relevant - use of SQLite and fact that pagination is computed "Server-side" are implementation details.

From an engineering perspective, we are changing the cache implementation to enable better pagination. Both aspects are equally important and have to go hand in hand. I do not care which one is given more prominence.

How about just using

```suggestion
title: UI Server-Side Pagination
```

And clarifying later that is achieved via a new SQLite cache?

Reviewer: Sounds good!

---

<head>
  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching"/>
</head>

:::caution
Caching Kubernetes objects with SQLite is not intended for use in production. This feature is considered highly experimental. SUSE customers should consult SUSE Support before activating this feature.
:::

> SQLite-backed caching provides an optional cache of Kubernetes objects to improve performance, especially while browsing tabular data with thousands of rows. The cache improves UI sorting, filtering, and pagination.

Reviewer: ```suggestion
SQLite-backed caching provides an optional cache of Kubernetes objects to improve performance. This unlocks sorting, filtering, and pagination features used by the UI to restrict the amount of resources it fetches and stores locally. These features are primarily used to improve list performance for resources with high counts by utilizing server-side pagination.
```

This feature creates file system based caches in the `rancher` pods of the upstream cluster, and in the `cattle-cluster-agent` pods of the downstream clusters. In most environments, disk usage and I/O should not be significant. However, you should monitor activity after you enable caching.

SQLite-backed caching persists copies of any cached Kubernetes objects to disk. See [Encrypting SQLite-backed Caching](#encrypting-sqlite-backed-caches) if this is a security concern.

## Enabling SQLite-Backed Caching

1. In the upper left corner, click **☰ > Global Settings > Feature Flags**.
1. Find **`ui-sql-cache`** and select **⋮ > Activate > Activate**.
1. Wait for Rancher to restart. This also restarts agents on all downstream clusters.
1. In the upper left corner, click **☰ > Global Settings > Performance**.

> 1. Go to **Server-side Pagination** and check the **Server-side Pagination** option.

Reviewer: ```suggestion
1. Go to **Server-side Pagination** and check the **Enable Server-side Pagination** option.
```

1. Click **Apply**.
1. Reload the page with the browser button (or the equivalent keyboard combination, typically `CTRL + R` on Windows and Linux, and `⌘ + R` on macOS).

## Encrypting SQLite-backed Caches

SQLite-backed caching persists copies of any cached Kubernetes objects to disk. If you're concerned about the safety of this data, you can encrypt all objects  before they are persisted to disk, by setting the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in the downstream clusters.

Secrets and security Tokens are always encrypted regardless of the above setting.

## Known Limitations of SQLite-Backed Caching

> Only Pods, Secrets, Node, ConfigMap and Event pages make use of the cache.

Reviewer: ```suggestion
This initial release improves the performance of Pods, Secrets, Nodes and ConfigMaps in the Cluster Explorer pages, and most resources in the Explorer's `More Resources` section
```

Reviewer: This second part of this list covers the other resources that make use of the cache

---

> title: SQLite-backed caching

Reviewer 2: ```suggestion
title: SQLite-Backed Caching
```

---

<head>
  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/how-to-guides/advanced-user-guides/enable-experimental-features/sqlite-caching"/>
</head>

:::caution

> Enabling the SQLite-backed caching of Kubernetes objects is a **highly experimental** feature, it is not yet officially supported in Rancher. It is not intended for any production use. SUSE customers should consult SUSE Support before activating this feature.

Reviewer 2: ```suggestion
Caching Kubernetes objects with SQLite is not intended for use in production. This feature is considered highly experimental. SUSE customers should consult SUSE Support before activating this feature.
```

:::

> ## About this Feature

Reviewer 2: ```suggestion
```

> This feature enables an optional cache of Kubernetes objects with the aim of improving performance, especially while browsing tabular data in the UI with thousands of rows. It also provides improved UI sorting, filtering and pagination features. 

Reviewer 2: ```suggestion
SQLite-backed caching provides an optional cache of Kubernetes objects to improve performance, especially while browsing tabular data with thousands of rows. The cache improves UI sorting, filtering, and pagination. 

This feature creates file system based caches in the `rancher` pods of the upstream cluster, and in the `cattle-cluster-agent` pods of the downstream clusters. In most environments, disk usage and I/O should not be significant. However, you should monitor activity after you enable caching.

SQLite-backed caching persists copies of any cached Kubernetes objects to disk. See [Encrypting SQLite-backed Caching](#encrypting-sqlite-backed-caches) if this is a security concern.
```

> ## Enabling instructions
> To enable the feature, perform the following steps:
> 1. Log into the Rancher UI.
> 1. In the upper left corner, click **☰ > Global Settings > Feature Flags**.
> 1. Go to **`ui-sql-cache`** and click **⋮ > Activate > Activate**.
> 1. Wait for Rancher to restart. This also restarts agents on all downstream clusters.
> 1. In the upper left corner, click **☰ > Global Settings > Performance**.
> 1. Go to **Server-side Pagination** and check the **Server-side Pagination** option.
> 1. Click **Apply**
> 1. Refresh the page (`CTRL + R`)

Reviewer 2: ```suggestion
## Enabling SQLite-Backed Caching

1. In the upper left corner, click **☰ > Global Settings > Feature Flags**.
1. Find **`ui-sql-cache`** and select **⋮ > Activate > Activate**.
1. Wait for Rancher to restart. This also restarts agents on all downstream clusters.
1. In the upper left corner, click **☰ > Global Settings > Performance**.
1. Go to **Server-side Pagination** and check the **Server-side Pagination** option.
1. Click **Apply**.
1. Refresh the page (`CTRL + R`).
```

Is there no button to refresh the page? If not, different OS's have different combinations of key presses to perform a refresh and I guess we'd have to list them out

Author: I addressed the refresh hotkey question in afe50d3d2a5

> ## Operational notes
> This feature creates file system based caches in the `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in downstream caches. In most use cases disk usage and I/O should not be significant, however monitoring is recommended.

Reviewer 2: ```suggestion
```

Moved into intro and removed header

> ## Security notes
> This feature persists copies of any cached Kubernetes objects to disk. If this is a security concern, an option exists to have all objects encrypted before being persisted to disk.
> 
> Set the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in the pods that need full encryption (`rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in downstream caches respectively).
> 
> Secrets and security Tokens are always encrypted regardless of the above setting.

Reviewer 2: ```suggestion
## Encrypting SQLite-backed Caches

SQLite-backed caching persists copies of any cached Kubernetes objects to disk. If you're concerned about the safety of this data, you can encrypt all objects  before they are persisted to disk, by setting the environment variable `CATTLE_ENCRYPT_CACHE_ALL` to `true` in `rancher` pods in the upstream cluster and `cattle-cluster-agent` pods in the downstream clusters.

Secrets and security Tokens are always encrypted regardless of the above setting.
```

> ## Known limitations
> At the time of writing the feature is subject to limitations below:
> 
> 1. Only the Pods, Secrets, Node, ConfigMap and Event pages make use of the cache.
> 1. Automatic refresh of pages is disabled. Table contents can be manually refreshed via a **Refresh** button.

Reviewer 2: ```suggestion
## Known Limitations of SQLite-Backed Caching

Only Pods, Secrets, Node, ConfigMap and Event pages make use of the cache.

Pages can't be automatically refreshed. You can manually refresh table contents by clicking the **Refresh** button.
```
